Should be able to derive `reflectapi::{Output, Input}` without extending type bound for other derives.

# What I Expected to Work

I would have expected the following code to compile without problems, similar to other derives like `Default` and `Debug` in the Rust ecosystem.

```rust
#[derive(Debug, serde::Serialize, reflectapi::Output)]
pub struct Items<T> {
    pub items: Vec<T>,
}
```

# What Happened
```shell
error[E0277]: the trait bound `T: reflectapi::Output` is not satisfied
 --> packages/backend/apps/core-server/libs/lookup/src/model/items.rs:4:18
  |
4 | pub struct Items<T> {
  |                  ^ the trait `reflectapi::Output` is not implemented for `T`
  |
help: consider restricting type parameter `T`
  |
4 | pub struct Items<T: reflectapi::Output> {
  |                   ++++++++++++++++++++

error[E0277]: the trait bound `T: reflectapi::Output` is not satisfied
 --> packages/backend/apps/core-server/libs/lookup/src/model/items.rs:5:16
  |
5 |     pub items: Vec<T>,
  |                ^^^^^^ the trait `reflectapi::Output` is not implemented for `T`, which is required by `Vec<T>: reflectapi::Output`
  |
  = note: required for `Vec<T>` to implement `reflectapi::Output`
help: consider restricting type parameter `T`
  |
4 | pub struct Items<T: reflectapi::Output> {
  |                   ++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `lookup` (lib) due to 2 previous errors
```

# Why is this bad?

Being forced to introduce this type bound is not common practice, and pollutes other types, e.g. it means that manually implementing other traits requires the bound too:
```rust
impl<T: ::std::clone::Clone> ::std::clone::Clone for Items<T>
where
    // XXX: This bound should not be required.
    T: reflectapi::Output,
{
    fn clone(&self) -> Self {
        Items {
            items: self.items.clone(),
        }
    }
}
```

The generated macro code calls T::reflect_type(). It can not be done
without such a constraint.

On Mon, 29 Jul 2024, 10:56 Josiah Bull, ***@***.***> wrote:

> What I Expected to Work
>
> I would have expected the following code to compile without problems,
> similar to other derives like Default and Debug in the Rust ecosystem.
>
> #[derive(Debug, serde::Serialize, reflectapi::Output)]pub struct Items<T> {
>     pub items: Vec<T>,}
>
> What Happened
>
> error[E0277]: the trait bound `T: reflectapi::Output` is not satisfied
>  --> packages/backend/apps/core-server/libs/lookup/src/model/items.rs:4:18
>   |
> 4 | pub struct Items<T> {
>   |                  ^ the trait `reflectapi::Output` is not implemented for `T`
>   |
> help: consider restricting type parameter `T`
>   |
> 4 | pub struct Items<T: reflectapi::Output> {
>   |                   ++++++++++++++++++++
>
> error[E0277]: the trait bound `T: reflectapi::Output` is not satisfied
>  --> packages/backend/apps/core-server/libs/lookup/src/model/items.rs:5:16
>   |
> 5 |     pub items: Vec<T>,
>   |                ^^^^^^ the trait `reflectapi::Output` is not implemented for `T`, which is required by `Vec<T>: reflectapi::Output`
>   |
>   = note: required for `Vec<T>` to implement `reflectapi::Output`
> help: consider restricting type parameter `T`
>   |
> 4 | pub struct Items<T: reflectapi::Output> {
>   |                   ++++++++++++++++++++
>
> For more information about this error, try `rustc --explain E0277`.
> error: could not compile `lookup` (lib) due to 2 previous errors
>
> Why is this bad?
>
> Being forced to introduce this type bound is not common practice, and
> pollutes other types, e.g. it means that manually implementing other traits
> requires the bound too:
>
> impl<T: ::std::clone::Clone> ::std::clone::Clone for Items<T>where
>     // XXX: This bound should not be required.
>     T: reflectapi::Output,{
>     fn clone(&self) -> Self {
>         Items {
>             items: self.items.clone(),
>         }
>     }}
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/thepartly/reflectapi/issues/11>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA6JSVKALYQQAD5PDRI466LZOVZKJAVCNFSM6AAAAABLTFWV22VHI2DSMVQWIX3LMV43ASLTON2WKOZSGQZTIMJUHE3TENY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/reflectapi-demo/tests/errors/invalid_reflect_type_generic.stderr
+++ b/reflectapi-demo/tests/errors/invalid_reflect_type_generic.stderr
@@ -6,25 +6,3 @@ error[E0392]: parameter `T` is never used
   |
   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
   = help: if you intended `T` to be a const parameter, use `const T: usize` instead
-
-error[E0277]: the trait bound `T: Input` is not satisfied
- --> tests/errors/invalid_reflect_type_generic.rs:2:17
-  |
-2 | struct MyStruct<T> {
-  |                 ^ the trait `Input` is not implemented for `T`
-  |
-help: consider restricting type parameter `T`
-  |
-2 | struct MyStruct<T: reflectapi::Input> {
-  |                  +++++++++++++++++++
-
-error[E0277]: the trait bound `T: reflectapi::Output` is not satisfied
- --> tests/errors/invalid_reflect_type_generic.rs:2:17
-  |
-2 | struct MyStruct<T> {
-  |                 ^ the trait `reflectapi::Output` is not implemented for `T`
-  |
-help: consider restricting type parameter `T`
-  |
-2 | struct MyStruct<T: reflectapi::Output> {
-  |                  ++++++++++++++++++++

--- a/reflectapi-demo/tests/success/reflect_generic_struct.rs
+++ b/reflectapi-demo/tests/success/reflect_generic_struct.rs
@@ -1,0 +1,4 @@
+#[derive(Debug, serde::Serialize, reflectapi::Input, reflectapi::Output)]
+pub struct Gen<T>(T);
+
+fn main() {}

--- a/reflectapi-derive/src/derive.rs
+++ b/reflectapi-derive/src/derive.rs
@@ -14,8 +14,16 @@ use crate::{
 pub(crate) fn derive_reflect(input: TokenStream, reflectapi_type: ReflectType) -> TokenStream {
     let input = syn::parse_macro_input!(input as syn::DeriveInput);
     let type_ident = input.ident.clone();
-    let type_generics = input.generics.clone();
-    let type_generics_where = input.generics.where_clause.clone();
+    let mut type_generics = input.generics.clone();
+
+    // Bound each generic by `reflectapi::Input` or `reflectapi::Output`.
+    for param in type_generics.type_params_mut() {
+        param.bounds.push(match reflectapi_type {
+            ReflectType::Input => parse_quote!(reflectapi::Input),
+            ReflectType::Output => parse_quote!(reflectapi::Output),
+        });
+    }
+    let type_generics_where = &input.generics.where_clause;
 
     let serde_context = serde_derive_internals::Ctxt::new();
     let serde_type_def = serde_derive_internals::ast::Container::from_ast(

